File: coldatomlab/echo.py
"""Independent spectral reference for paired isolated holds and a midpoint mode swap."""

import math

import numpy as np

from . import preparation
from .twomode import Reference, _compare, diagnostics

CONVENTION = (
    "J=0; H/h=U(n-N/2)^2-bias(n-N/2); midpoint S|n,N-n>=|N-n,n>; fixed-N pulse global phase omitted"
)


def validate(plan):
    p = preparation.validate(plan)
    if p["base"]["tunnelling_hz"] != 0:
        raise ValueError("Spin echo requires J=0 during both holds.")
    return p


def checkpoints(duration):
    times = [duration * i / 100 for i in range(101)]
    times.insert(51, duration / 2)
    return [dict(time_ms=t, pulse_applied=i >= 51) for i, t in enumerate(times)]


def evolve(recipe):
    ref = Reference(recipe["config"])
    c = ref.config
    if c["tunnelling_hz"] != 0:
        raise ValueError("Spin echo requires J=0.")
    v, energies = ref.vectors, ref.eigenvalues
    initial = v @ ref.coefficients
    half = c["duration_ms"] / 2

    def advance(state, time):
        return v @ (np.exp(-2j * np.pi * energies * time / 1000) * (v.T @ state))

    def observe(state, time):
        return dict(
            time_ms=time,
            real=state.real.tolist(),
            imag=state.imag.tolist(),
            **diagnostics(c, state),
        )

    before = advance(initial, half)
    swap_operator = np.eye(c["atoms"] + 1)[::-1]
    after = swap_operator @ before
    record = dict(
        **recipe,
        pulse_before=observe(before, half),
        pulse_after=observe(after, half),
        no_echo=dict(history=[]),
        echo=dict(history=[]),
    )
    for point in checkpoints(c["duration_ms"]):
        for arm in ("no_echo", "echo"):
            applied = arm == "echo" and point["pulse_applied"]
            time = point["time_ms"]
            psi = advance(after if applied else initial, time - half if applied else time)
            state = observe(psi, time)
            record[arm]["state"] = state
            record[arm]["history"].append(
                dict(**{k: state[k] for k in preparation.HISTORY_KEYS}, pulse_applied=applied)
            )
    return record


def ideal(plan):
    return evolve(dict(index=-1, phase_offset=0, bias_offset_hz=0, config=plan["base"]))


def aggregate(plan, reference, records):
    if not records:
        return None
    result = {}
    for arm in ("no_echo", "echo"):
        result[arm] = preparation.aggregate(plan, reference[arm], [r[arm] for r in records])
        for i, row in enumerate(result[arm]["history"]):
            row["pulse_applied"] = reference[arm]["history"][i]["pulse_applied"]
            effective = row["time_ms"] - (
                plan["base"]["duration_ms"] if row["pulse_applied"] else 0
            )
            row["guide_coherence"] = float(
                row["ideal_coherence"]
                * abs(
                    np.sinc(plan["phase_half_range"] / np.pi)
                    * np.sinc(2 * plan["bias_half_range_hz"] * effective / 1000)
                )
            )
    return result


def compare_tree(saved, expected, context="Echo"):
    """Strict structure plus the established absolute/phase-aware numeric tolerance."""
    if isinstance(expected, dict):
        if not isinstance(saved, dict) or set(saved) != set(expected):
            raise ValueError(f"{context}: fields differ.")
        for key, value in expected.items():
            if key == "phase":
                _compare({key: saved[key]}, {key: value}, context)
            else:
                compare_tree(saved[key], value, f"{context}.{key}")
    elif isinstance(expected, list):
        if not isinstance(saved, list) or len(saved) != len(expected):
            raise ValueError(f"{context}: lengths differ.")
        for i, (a, b) in enumerate(zip(saved, expected, strict=True)):
            compare_tree(a, b, f"{context}[{i}]")
    elif isinstance(expected, (str, bool)) or expected is None:
        if type(saved) is not type(expected) or saved != expected:
            raise ValueError(f"{context}: value differs.")
    elif (
        isinstance(saved, bool)
        or not isinstance(saved, (float, int))
        or not math.isfinite(saved)
        or abs(saved - expected) > 5e-8
    ):
        raise ValueError(f"{context}: value differs.")


def verify_export(data):
    if data.get("schema") != "coldatomlab-echo-v1" or data.get("convention") != CONVENTION:
        raise ValueError("Unknown echo convention.")
    plan = validate(data["plan"])
    count = len(data["records"])
    if data["status"] not in ("complete", "cancelled") or not 1 <= count <= plan["preparations"]:
        raise ValueError("Invalid completed-prefix status.")
    if data["status"] == "complete" and count != plan["preparations"]:
        raise ValueError("Incomplete comparison marked complete.")
    reference = ideal(plan)
    records = [evolve(recipe) for recipe in preparation.recipes(plan)[:count]]
    compare_tree(data["ideal"], reference)
    compare_tree(data["records"], records)
    compare_tree(data["aggregate"], aggregate(plan, reference, records))
    errors, drift = [], []
    for saved, expected in zip(
        [data["ideal"], *data["records"]], [reference, *records], strict=True
    ):
        for a, b in (
            (saved["pulse_before"], expected["pulse_before"]),
            (saved["pulse_after"], expected["pulse_after"]),
            (saved["no_echo"]["state"], expected["no_echo"]["state"]),
            (saved["echo"]["state"], expected["echo"]["state"]),
        ):
            psi = np.asarray(a["real"]) + 1j * np.asarray(a["imag"])
            oracle = np.asarray(b["real"]) + 1j * np.asarray(b["imag"])
            errors.append(float(np.linalg.norm(psi - oracle)))
            drift.append(abs(a["norm"] - 1))
        for arm in ("no_echo", "echo"):
            drift.extend(abs(row["norm"] - 1) for row in saved[arm]["history"])
    if max(errors) > 5e-8 or max(drift) > 1e-9:
        raise ValueError("Echo state or norm tolerance exceeded.")
    return dict(
        verified=True,
        paired_preparations_verified=count,
        status=data["status"],
        maximum_complex_state_l2=max(errors),
        maximum_norm_drift=max(drift),
        history_rows_per_arm=102,
        reference="Independent NumPy spectral propagation and explicit mode-swap matrix",
    )

File: coldatomlab/web/echo.js
"use strict";
// Exact isolated-well evolution; the instantaneous pulse swaps occupation amplitudes.
(() => {
  const convention = "J=0; H/h=U(n-N/2)^2-bias(n-N/2); midpoint S|n,N-n>=|N-n,n>; fixed-N pulse global phase omitted";
  const defaults = { ...Preparation.defaults, base: { ...Preparation.defaults.base }, phase_half_range: 0, bias_half_range_hz: 2 };
  function validate(input) {
    const p = Preparation.validate(input);
    if (p.base.tunnelling_hz !== 0) throw new Error("Spin echo requires J = 0 during the holds. Nonzero tunnelling is not silently removed.");
    return p;
  }
  function checkpoints(duration) {
    const rows = [];
    for (let i = 0; i <= 100; i++) {
      rows.push({ time_ms: duration * i / 100, pulse_applied: i > 50 });
      if (i === 50) rows.push({ time_ms: duration / 2, pulse_applied: true });
    }
    return rows;
  }
  function swap(state) { return { r: Float64Array.from(state.r).reverse(), im: Float64Array.from(state.im).reverse() }; }
  function propagate(c, state, time_ms) {
    if (c.tunnelling_hz !== 0 || !Number.isFinite(time_ms) || time_ms < 0) throw new Error("Invalid isolated hold.");
    const r = new Float64Array(c.atoms + 1), im = new Float64Array(c.atoms + 1);
    for (let k = 0; k <= c.atoms; k++) {
      const m = k - c.atoms / 2, a = -2 * Math.PI * (c.interaction_hz * m * m - c.bias_hz * m) * time_ms / 1000;
      r[k] = state.r[k] * Math.cos(a) - state.im[k] * Math.sin(a);
      im[k] = state.r[k] * Math.sin(a) + state.im[k] * Math.cos(a);
    }
    return { r, im };
  }
  function observe(c, state, time_ms) {
    const d = TwoMode.diagnostics(c, state.r, state.im);
    if (!Number.isFinite(d.norm) || Math.abs(d.norm - 1) > 1e-9) throw new Error("Echo norm check failed; no renormalization applied.");
    return { time_ms, real: Array.from(state.r), imag: Array.from(state.im), ...d };
  }
  function evolve(recipe) {
    const c = TwoMode.validate(recipe.config);
    if (c.tunnelling_hz !== 0) throw new Error("Spin echo requires J = 0.");
    const initial = TwoMode.initial(c), half = c.duration_ms / 2;
    const before = propagate(c, initial, half), after = swap(before);
    const record = { ...recipe, pulse_before: observe(c, before, half), pulse_after: observe(c, after, half), no_echo: { history: [] }, echo: { history: [] } };
    for (const point of checkpoints(c.duration_ms)) {
      for (const arm of ["no_echo", "echo"]) {
        const applied = arm === "echo" && point.pulse_applied;
        const state = observe(c, propagate(c, applied ? after : initial, applied ? point.time_ms - half : point.time_ms), point.time_ms);
        record[arm].state = state;
        record[arm].history.push({ ...Object.fromEntries(Preparation.historyKeys.map(k => [k, state[k]])), pulse_applied: applied });
      }
    }
    return record;
  }
  function ideal(plan) { return evolve({ index: -1, phase_offset: 0, bias_offset_hz: 0, config: { ...plan.base } }); }
  function aggregate(plan, reference, records) {
    if (!records.length) return null;
    return Object.fromEntries(["no_echo", "echo"].map(arm => {
      const result = Preparation.aggregate(plan, reference[arm], records.map(r => r[arm]));
      result.history.forEach((row, i) => {
        row.pulse_applied = reference[arm].history[i].pulse_applied;
        const effective = row.pulse_applied ? row.time_ms - plan.base.duration_ms : row.time_ms;
        row.guide_coherence = row.ideal_coherence * Math.abs(Preparation.sinc(plan.phase_half_range) * Preparation.sinc(2 * Math.PI * plan.bias_half_range_hz * effective / 1000));
      });
      return [arm, result];
    }));
  }
  function report(plan, reference, records, status) {
    return { schema: "coldatomlab-echo-v1", version: "0.14.0", convention, plan, status, ideal: reference, records, aggregate: aggregate(plan, reference, records) };
  }
  globalThis.Echo = { defaults, convention, validate, checkpoints, swap, propagate, observe, evolve, ideal, aggregate, report };
})();
